File: cakk-common/src/main/java/com/cakk/common/response/ApiResponse.java
package com.cakk.common.response;

import lombok.Getter;

import com.cakk.common.enums.ReturnCode;

@Getter
public class ApiResponse<T> {

	private String returnCode;
	private String returnMessage;
	private T data;

	public static <T> ApiResponse<T> success(T data) {
		ApiResponse<T> response = new ApiResponse<>();

		response.returnCode = ReturnCode.SUCCESS.getCode();
		response.returnMessage = ReturnCode.SUCCESS.getMessage();
		response.data = data;

		return response;
	}

	public static <T> ApiResponse<T> fail(ReturnCode returnCode, T data) {
		ApiResponse<T> response = new ApiResponse<>();

		response.returnCode = returnCode.getCode();
		response.returnMessage = returnCode.getMessage();
		response.data = data;

		return response;
	}

	public static ApiResponse<String> error(ReturnCode returnCode, String errorMessage) {
		ApiResponse<String> response = new ApiResponse<>();

		response.returnCode = returnCode.getCode();
		response.returnMessage = returnCode.getMessage();
		response.data = errorMessage;

		return response;
	}
}

File: cakk-common/src/main/kotlin/com/cakk/common/response/ApiResponse.kt
package com.cakk.common.response

import com.cakk.common.enums.ReturnCode

data class ApiResponse<T>(
	val returnCode: String,
	val returnMessage: String,
	val data: T? = null
) {

	constructor(returnCode: ReturnCode, data: T? = null) : this(returnCode.code, returnCode.message, data)

	companion object {

		fun <T> success(data: T? = null): ApiResponse<T> {
			return ApiResponse(
				returnCode = ReturnCode.SUCCESS,
				data = data
			)
		}

		fun <T> fail(returnCode: ReturnCode, data: T? = null): ApiResponse<T> {
			return ApiResponse(
				returnCode = returnCode,
				data = data
			)
		}

		fun error(returnCode: ReturnCode, errorMessage: String?): ApiResponse<String> {
			return ApiResponse(
				returnCode = returnCode,
				data = errorMessage
			)
		}
	}
}
